### ISSUE TITLE
<x-dynamic-component /> support?

### ISSUE THREAD

**Issue by stevenmaguire** [tighten/jigsaw#707]
Is the Laravel documented support for [Dynamic Components](https://laravel.com/docs/11.x/blade#dynamic-components) is expected to work in Jigsaw?

An exception is raised when trying it out and it's unclear if it is local to the project or broadly unsupported in Jigsaw.

```php
$componentType = 'something-kebab-case';

<x-dynamic-component :component="$componentType" />
```

```bash
   Spatie\LaravelIgnition\Exceptions\ViewException 
  Target [Illuminate\Contracts\Foundation\Application] is not instantiable.
```

In this example, a component is confirmed as working when expressed explicitly, so it is not a missing component issue.

This works just fine.

```php
<x-something-kebab-case />
```

The error _feels_ architectural. 

I also noticed there is some reference to dynamic component support in the Jigsaw providers, but there are also some code comments and TODOs. 

https://github.com/tighten/jigsaw/blob/114ed1650af48ea9c4ce333d08cc45c97ad5b928/src/Providers/ViewServiceProvider.php#L61-L72

Is this expected to be supported and unfinished?

**Comment by stevenmaguire** [comment]
Digging into this further, there is definitely something incomplete here in Jigsaw. I'll try to explain what that might be and I think that Tighten may want to decide if this functionality is actually meant to be supported. 

tl;dr - it seems to be incomplete. hopefully my observations will help to narrow the scope of what can be done to add support, if that is important.

------------------------------- 

The main seam where this support is falling apart is in the compiler method that is attempting to guess the component name, view path, and/or class name. 

```php
Illuminate\View\Compilers\ComponentTagCompiler::componentClass(string $component);
```

<img width="808" alt="CleanShot 2024-06-14 at 15 39 46@2x" src="https://github.com/tighten/jigsaw/assets/1851973/23e847eb-84a0-45c4-b196-8d4540da6013">

The exception initially reported was cause by a `make` call on the container to resolve a class that is bound to the `Illuminate\Contracts\Foundation\Application` interface; the method responsible for that is `Illuminate\View\Compilers\ComponentTagCompiler::guessClassName`.

https://github.com/laravel/framework/blob/6c0561caeb8265b7d17cabeea0ec7d4c5cca28aa/src/Illuminate/View/Compilers/ComponentTagCompiler.php#L409-L418

I added the following to my `bootstrap.php` file to get past the resolution issue. 

```php
$events->beforeBuild([
    fn ($jigsaw) => $jigsaw->app->bind(\Illuminate\Contracts\Foundation\Application::class, fn () => $jigsaw->app)
]);
```

`$jigsaw->app`, which is an instance of `TightenCo\Jigsaw\Container` does not currently implement that interface (perhaps that could|would be a good idea?) and that doesn't matter. What does matter, and for this very specific use case, is that it furnishes a `getNamespace` method that returns a string. The returned string is expected to be used to fill out a Laravel-specific namespace to find a component - this will always fail. So, it can return an empty string.

That is not going to produce a value for `$class` so the logic will move on to `guessAnonymousComponentUsingNamespaces` and `guessAnonymousComponentUsingPaths`.

`guessAnonymousComponentUsingNamespaces` is likely going to always fail, even if you attempt to register the local `\Components` namespace because there is no actual namespace. `guessAnonymousComponentUsingPaths` _feels_ like the right place to focus effort to resolve a good component name. 

Although, at the moment, even attempting to register a path in `blade.php`, there is yet another missing container binding for `views`, just like the Application interface binding mentioned above.

```php
$events->beforeBuild([
    fn ($jigsaw) => $jigsaw->app->bind(\Illuminate\Contracts\View\Factory::class, fn () => $jigsaw->app->make('view'))
]);
```

```php
$bladeCompiler->anonymousComponentPath(__DIR__.'/source/_components');
```

This still fails due to some config issue, also with TODO code comments in `vendor/tightenco/jigsaw/src/Providers/ViewServiceProvider.php:57`.

<img width="826" alt="CleanShot 2024-06-14 at 15 49 55@2x" src="https://github.com/tighten/jigsaw/assets/1851973/dd2a77b2-3206-44f6-bb55-cccc6f0e0ee6">

**Comment by stevenmaguire** [comment]
Sorry for the mention @damiani but I am curious if the Tighten team has an opinion on this? Is this an incomplete expected feature or something that should not be expected?

**Comment by damiani** [comment]
@stevenmaguire Thanks for the reminder ... I'll take a look and see what we'll need to do to support this; appreciate your detailed analysis of the code paths involved. More shortly!

**Comment by stevenmaguire** [comment]
@damiani thanks for the prompt response. Sorry again to escalate - it seems like an oversight one way or another.

**Comment by stevenmaguire** [comment]
> It looks like I added this in https://github.com/tighten/jigsaw/pull/662 as part of a much larger refactor, but I'll keep staring at it and see if I remember anything.

> Actually we're using our own ComponentTagCompiler to do some custom path/namespace resolution stuff, I wonder if we just need to use our own DynamicComponent too so that we can force it to use our compiler instead of Laravel's.

These two statements feel more relevant than the first question asking about the folder structure. I am highly confident it is not simply a case of pathing and/or directory issues. After I stepped through it all, I am firmly in the camp of "it feels like an architectural or plumbing issue." Using your own compiler would likely resolve it and force you to touch each of the seams I identified in my second comment here.
